datastream-kafka/src/test/java/com/linkedin/datastream/kafka/TestKafkaProducerWrapper.java

```diff
@@ -141,8 +153,17 @@ void verifyFlush(int numExpected) {
       verify(_mockProducer, times(numExpected)).flush();
     }
 
-    void verifyClose(int numExpected) {
+    void verifyClose(int numExpected) throws NoSuchMethodException {
+      // Producer close is invoked in a separate thread. Must wait for the thread to get scheduled and call close
+      Method method = Producer.class.getMethod("close", long.class, TimeUnit.class);
+      PollUtils.poll(() -> {
+        Collection<Invocation> invocations = mockingDetails(_mockProducer).getInvocations();
+        long count = invocations.stream().filter(invocation -> invocation.getMethod().equals(method)).count();
+        return count == numExpected;
+      }, 1000, 10000);
       verify(_mockProducer, times(numExpected)).close(anyLong(), any(TimeUnit.class));
```

Comment:
Please, feel free to ignore this comment.

I generally dislike polling in tests for several reasons. I'll just mention the most relevant ones here:
1. It's a source of flakiness (I know you used a timeout of 10s for that reason)
2. When we do `verifyClose(0)`, `count == numExpected` will return `true` immediately, a racey behavior that does not affect this test because it has a subsequent assertion that verifies `shutdownProducer()` itself hasn't been called.
    ```java
    Assert.assertEquals(_numShutdownProducerCalls, numExpected)
    ```
    but this begs the question: do we have to assert on `_mockProducer.produce()` in the first place?

Reply:
As discussed offline, it does make sense to leave this as is since we want to test both that shutdownProducer is called and that mockProducer's close() is called.